lib/backend/k8s/syncer.go

```diff
@@ -0,0 +1,536 @@
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	k8sapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/watch"
+	"reflect"
+)
+
+func newSyncer(kc KubeClient, callbacks api.SyncerCallbacks) *kubeSyncer {
+	return &kubeSyncer{
+		kc:        kc,
+		callbacks: callbacks,
+	}
+}
+
+type kubeSyncer struct {
+	kc        KubeClient
+	callbacks api.SyncerCallbacks
+	OneShot   bool
+}
+
+// Holds resource version information.
+type resourceVersions struct {
+	podVersion           string
+	namespaceVersion     string
+	networkPolicyVersion string
+}
+
+func getUpdateType(e watch.Event) model.UpdateType {
+	// Determine the update type.
+	updateType := model.UpdateTypeKVUnknown
+	switch e.Type {
+	case watch.Added:
+		updateType = model.UpdateTypeKVNew
+	case watch.Deleted:
+		updateType = model.UpdateTypeKVDeleted
+	case watch.Modified:
+		updateType = model.UpdateTypeKVUpdated
+	}
+	return updateType
+}
+
+func (syn *kubeSyncer) Start() {
+	// Channel for receiving updates from a snapshot.
+	snapshotUpdates := make(chan *[]model.Update)
+
+	// Channel for receiving updates from the watcher.
+	watchUpdates := make(chan *model.Update)
+
+	// Channel used by the watcher to trigger a re-sync.
+	triggerResync := make(chan *resourceVersions, 5)
+
+	// Channel to send the index from which to start the snapshot.
+	initialSnapshotIndex := make(chan *resourceVersions)
+
+	// Status channel
+	statusUpdates := make(chan api.SyncStatus)
+
+	// If we're not in one-shot mode, start the API watcher to
+	// gather updates.
+	if !syn.OneShot {
+		go syn.watchKubeAPI(watchUpdates, triggerResync, initialSnapshotIndex)
+	}
+
+	// Start a background thread to read snapshots from etcd.  It will
+	// read a start-of-day snapshot and then wait to be signalled on the
+	// resyncIndex channel.
+	go syn.readSnapshot(snapshotUpdates, triggerResync, initialSnapshotIndex, statusUpdates)
+
+	// Start a routine to merge updates from the snapshot routine and the
+	// watch routine (if running), and pass information to callbacks.
+	// TODO: We're never actually running both the snapshot thread and the watch thread at the same time
+	// so what's the point?
+	go syn.mergeUpdates(snapshotUpdates, watchUpdates, statusUpdates)
+}
+
+// mergeUpdates ensures that callbacks are all executed from the same thread.
+func (syn *kubeSyncer) mergeUpdates(snapshotUpdates chan *[]model.Update, watchUpdates chan *model.Update, statusUpdates chan api.SyncStatus) {
+	var update *model.Update
+	var updates *[]model.Update
+	currentStatus := api.WaitForDatastore
+	newStatus := api.WaitForDatastore
+	syn.callbacks.OnStatusUpdated(api.WaitForDatastore)
+	for {
+		select {
+		case updates = <-snapshotUpdates:
+			log.Debugf("Snapshot update: %+v", updates)
+			if currentStatus != api.ResyncInProgress {
+				panic("Recieved snapshot update while not resyncing")
+			}
+			syn.callbacks.OnUpdates(*updates)
```

Comment:
I think you're missing handling of deletions here.  If you lose sync with k8s and then start a new snapshot, some pods may have been deleted in the meantime. You need to keep some sort of datastructure (say a set of all known Key objects) to spot that case.

You don't need a HWM tracker (a la etcd) because your resync algorithm is simpler.